Documentation update for reference resolution without expansion on qu… (#710)
commercetools/commercetools-sync-java: docs/usage/CATEGORY_SYNC.md

##### Syncing from a commercetools project

When syncing from a source commercetools project, you can use [`transformCategoryReferences`](https://commercetools.github.io/commercetools-sync-java/v/4.0.1/com/commercetools/sync/categories/service/CategoryReferenceTransformService.html#transformCategoryReferences-java.util.List-)

> the method that transforms(resolves by querying and caching key values for Ids) and maps from a `Category` to `CategoryDraft` using cache in order to make them ready for reference resolution by the sync, for example: 

Reviewer: ```suggestion method that transforms(resolves by querying and caching key values for Ids) and maps from a `Category` to `CategoryDraft` using cache in order to make them ready for reference resolution by the sync, for example: 
```

> the method that transforms(resolves by querying and caching key values for Ids) and maps from a `Category` to `CategoryDraft` using cache in order to make them ready for reference resolution by the sync, for example: 

Reviewer: ```suggestion
 method that transforms(resolves by querying and caching key-id pairs) and maps from a `Category` to `CategoryDraft` using cache in order to make them ready for reference resolution by the sync, for example: 
```

````java
// Build a CategoryQuery for fetching categories from a source CTP project without any references expanded for the sync:
final CategoryQuery categoryQueryWithReferences = CategoryReferenceResolutionUtils.buildCategoryQuery();

// Query all categories (NOTE this is just for example, please adjust your logic)
final List<Category> categories =
    CtpQueryUtils
        .queryAll(sphereClient, categoryQueryWithReferences, Function.identity())
        .thenApply(fetchedResources -> fetchedResources
            .stream()
            .flatMap(List::stream)
            .collect(Collectors.toList()))
        .toCompletableFuture()
        .join();
````

In order to transform and map the category, 
Initialize [`CategoryReferenceTransformService`](https://github.com/commercetools/commercetools-sync-java/tree/master/src/main/java/com/commercetools/sync/categories/service/CategoryReferenceTransformService.java) with `sphereClient` and cache(You can use your own cache implementation and pass the map).

> For cache implementation, you can refer an example class in the library - which implements the cache using caffeine library with an LRU based cache eviction strategy[`InMemoryReferenceIdToKeyCache`](https://github.com/commercetools/commercetools-sync-java/tree/master/src/main/java/com/commercetools/sync/commons/utils/InMemoryReferenceIdToKeyCache.java).

Reviewer: ```suggestion
For cache implementation, you can refer an example class in the library - which implements the cache using caffeine library with an LRU(Last Recently Used)-based cache eviction strategy[`InMemoryReferenceIdToKeyCache`](https://github.com/commercetools/commercetools-sync-java/tree/master/src/main/java/com/commercetools/sync/commons/utils/InMemoryReferenceIdToKeyCache.java).
```

> For cache implementation, you can refer an example class in the library - which implements the cache using caffeine library with an LRU based cache eviction strategy[`InMemoryReferenceIdToKeyCache`](https://github.com/commercetools/commercetools-sync-java/tree/master/src/main/java/com/commercetools/sync/commons/utils/InMemoryReferenceIdToKeyCache.java).

Reviewer: ```suggestion
For cache implementation, you can refer an example class in the library - which implements the cache using caffeine library with an LRU (Last Recently Used) based cache eviction strategy[`InMemoryReferenceIdToKeyCache`](https://github.com/commercetools/commercetools-sync-java/tree/master/src/main/java/com/commercetools/sync/commons/utils/InMemoryReferenceIdToKeyCache.java).
```

Author: already fixed so will resolve this conversation.

> For cache implementation, you can refer an example class in the library - which implements the cache using caffeine library with an LRU based cache eviction strategy[`InMemoryReferenceIdToKeyCache`](https://github.com/commercetools/commercetools-sync-java/tree/master/src/main/java/com/commercetools/sync/commons/utils/InMemoryReferenceIdToKeyCache.java).

Reviewer: ```suggestion
For cache implementation, you can refer an example class in the library - which implements the cache using caffeine library with an LRU (Last Recently Used) based cache eviction strategy[`InMemoryReferenceIdToKeyCache`](https://github.com/commercetools/commercetools-sync-java/tree/master/src/main/java/com/commercetools/sync/commons/utils/InMemoryReferenceIdToKeyCache.java).
```

Author: Already resolved so will close this conversation.

> Then call the transformCategoryReferences method with the `categories` parameter as shown below:

Reviewer: ```suggestion
Then call the `transformCategoryReferences` method with the `categories` parameter as shown below:
```

````java
// Fetch(Id to key values for references) into the cache and map from Category to CategoryDraft using cache with considering reference resolution.
final List<CategoryDraft> categoryDrafts = CategoryReferenceTransformService.transformCategoryReferences(categories);
````

The cache here is used for a better performance. 

> Instead of expanding the references in the query for category resource. `CategoryReferenceTransformService` will execute a query to fetch key values for the Ids(nonCached) and store in cache. These cached id to key values then can be used by another resource for resolving its references instead of fetching from commercetools API. It turns out, having the in-memory LRU cache will improve the overall performance of the sync library and commercetools API.

Reviewer: ```suggestion
Instead of expanding the references in the query for category resource. `CategoryReferenceTransformService` will execute a query to fetch key-id pairs and store in cache. These cached id to key values then can be used by another resource for resolving its references instead of fetching from commercetools API. It turns out, having the in-memory LRU cache will improve the overall performance of the sync library and commercetools API.
```

> Instead of expanding the references in the query for category resource. `CategoryReferenceTransformService` will execute a query to fetch key values for the Ids(nonCached) and store in cache. These cached id to key values then can be used by another resource for resolving its references instead of fetching from commercetools API. It turns out, having the in-memory LRU cache will improve the overall performance of the sync library and commercetools API.

Reviewer: ```suggestion
Instead of expanding the references in the query for category resource. `CategoryReferenceTransformService` will execute a query to fetch key-id pairs and store in cache. These cached id to key values then can be used by another resource for resolving its references instead of fetching from commercetools API. It turns out, having the in-memory LRU cache will improve the overall performance of the sync library and commercetools API.
```

Author: already fixed, So will resolve this conversation.

> Instead of expanding the references in the query for category resource. `CategoryReferenceTransformService` will execute a query to fetch key values for the Ids(nonCached) and store in cache. These cached id to key values then can be used by another resource for resolving its references instead of fetching from commercetools API. It turns out, having the in-memory LRU cache will improve the overall performance of the sync library and commercetools API.

Reviewer: ```suggestion
Instead of expanding the references in the query for category resource. `CategoryReferenceTransformService` will execute a query to fetch key-id pairs and store in cache. These cached id to key values then can be used by another resource for resolving its references instead of fetching from commercetools API. It turns out, having the in-memory LRU cache will improve the overall performance of the sync library and commercetools API.
```

Author: Already fixed, so will close this conversation.
